File: 백준 문제 풀이 모음/python/class 4/1916.py
import sys, heapq
INF = sys.maxsize

input = sys.stdin.readline

n = int(input())
m = int(input())

graph = [[] for i in range(n+1)]

for i in range(m) :
    a, b, c,  = map(int, input().split())
    graph[a].append((b, c))

start, end = map(int, input().split())


def dijkstra(start) :
    distance = [INF] * (n+1)
    distance[start] = 0
    heap = [(0, start)]
    
    while heap :
        now_cost, node = heapq.heappop(heap)
        if distance[node] < now_cost :
            continue

        for next_node, next_cost in graph[node] :
            next_cost  = next_cost + now_cost

            if distance[next_node] > next_cost :
                heapq.heappush(heap, (next_cost, next_node))
                distance[next_node] = next_cost


    return distance




arr = dijkstra(start)
print(arr[end])

File: 백준 문제 풀이 모음/kotlin/class 4/1916.kt
import java.util.PriorityQueue

private const val INF = Integer.MAX_VALUE / 2
private lateinit var dist : IntArray
private lateinit var list : MutableList<MutableList<Node>>

private class Node (var nodeNum : Int, var weight : Int) : Comparable<Node> {
    override fun compareTo(other: Node): Int {
        return weight - other.weight
    }

}

fun main() = with(System.`in`.bufferedReader()){
    val n = readLine().toInt()
    val m = readLine().toInt()

    dist = IntArray(n+1) { INF }
    list = ArrayList()
    repeat(n+1) {
        list.add(ArrayList())
    }

    repeat(m) {
        val input = readLine().split(' ').toList()
        val start = input[0].toInt()
        val end = input[1].toInt()
        val weight = input[2].toInt()

        list.get(start).add(Node(end, weight))
    }

    val str = readLine().split(' ').toList()
    val start = str[0].toInt()
    val end = str[1].toInt()

    println(dijkstra(start, end, n, m))

}

fun dijkstra(start: Int, end: Int, n:Int, m:Int): Int {
    val deque = PriorityQueue<Node>()
    val visited = MutableList<Boolean>(n+1) { false }
    dist[start] = 0
    deque.add(Node(start, 0))

    while (deque.isNotEmpty()) {
        val node = deque.poll()
        var start_node = node.nodeNum

        if (!visited[start_node]) {
            visited[start_node] = true

            list.get(start_node).forEach {
                if (!visited[it.nodeNum] && dist[it.nodeNum] > dist[start_node] + it.weight) {
                    dist[it.nodeNum] = dist[start_node] + it.weight
                    deque.offer(Node(it.nodeNum, dist[it.nodeNum]))
                }
            }
        }

    }
    return dist[end]
}
